DCP-3538: Improvement - Maintenance & refactor

KirillKirnichansky opened the pull request:
This MR resolves DCP-3538. The ticket originally was a maintenance task to make quality of life improvements, readme corrections and code cleanup as a result of the OAM MR. As the OAM MR developed, requirements for the worker hook updated and turned out a significant refactor was in order for the hook.

Originally the worker was not returned to prevent users/devs accessing/editing the worker internals and supervisor options the wrong way. This is not desired. The Worker API is public and supposed to be user facing, therefore, no reason to now extend the Worker API with this hook. There are custom API methods that just wrap worker API methods - this is not easily maintainable as any updates to the worker API will always need to update the hook API. Furthermore, workerOptions is currently in a useState. While this works, I want to mimic how the worker is used in our other worker applications: mutating the workerOptions object directly. Of course state is useful and needed for component updates/re-renders, however, this can be achieved using a Proxy. The Proxy will behave like a normal obj, except will trigger the necessary side-effects (re-render, saving to local storage).

Amongst the changes mentioned above, there are improved comments, error handling, type correction and general quality of life improvements.

These are breaking changes.

KirillKirnichansky committed:
improvement: replace dispatch action names w enum

reducer dispatch action names for both WorkerStateReducer and
WorkerStatsReducer have been put into an enum. This avoids future
problems in mispelling and mismatching action names

KirillKirnichansky committed:
style: use Allman brace style

KirillKirnichansky committed:
feature: add OAM type and return worker OAM

KirillKirnichansky committed:
improvement: change localStorage var name

There were changes to localStorage that altered which props are stored.
People's old localstorage vars were causing bugs, so I changed the name
so that moving forward, dcp-worker-options is the variable the hook
interacts with. The old var is ignored.

KirillKirnichansky committed:
improvement: big refactor caused by returning the worker

Originally the worker was not returned on purpose to prevent devs from accessing/editing
worker internals the wrong way. With improvements to the worker constructor, worker and supervisor, it turns out
we do want to return the worker. After all, dcp-worker is a public repo with a public API and documentation that is user facing. As a result, there have been some refactoring and improvements made to the hook:

- proper handling of race-condition in constructing the worker when/if the hook is executed multiple times at once
- applyWorkerOptions -> applyUserOptions
  - applyWorkerOptions no longer makes sense since editing workerOptions should be done by directly mutating workerOptions
  - moved all the applying of user specific options (hook params + local storage) to applyUserOptions
- added method to ensure paymentAddress is of a valid type
- worker options construction step is now its own method
- workerOptions passed to worker constructor is a Proxy now
  - needed a custom setter to reflect changes to paymentAddress and maxWorkingSandbox props to local storage
  - regardless if local storage is enabled, the setter is also used to trigger re-renders in components using the hook
  - paymentAddres and maxWorkingSandboxes are desired props to display in the front-end
- removed workerState.isLoaded since now we are returning the worker
  - !isLoaded === !worker
- removed workerOptionsState, start/stop/toggleWorker, sandboxes, originManager
  - these are all API methods of the worker
- improved error handling
- improved comments all through out
- quality of life improvements
- fixed types

KirillKirnichansky committed:
improvment: Update README to reflect refactor

Updated README once again to reflect changes since the last README
update

RobertMirandola commented on README.md:
```diff
@@ -56,14 +52,14 @@ function Worker() {
 - `identity?: Keystore`: A Keystore object (`dcp.wallet.Keystore`) which is passed to the Worker constructor and set as the Worker's identity when communicating over the network. If a Keystore is not provided, an arbitrary one will be generated.
 - `useLocalStorage?: boolean = true`:  A flag to toggle the use of the browser's local storage. The `workerOptions` object is the entity to be saved to local storage and is updated accordingly when calling `setWorkerOptions`.
-- `workerOptions: object`: This object is supplied to the Worker constructor as the `workerOptions` parameter (required). The only required property of the `workerOptions` object needed to provide is a `paymentAddress`. The rest of the properties will get default values.
+- `workerOptions: object`: The contents of this object will override the default values coming from the worker configuration provided by the `dcp-worker` library (is this correct?? dcp-client library? I know the web-config service is responsible for dcpConfig.worker). The only required property of the `workerOptions` object needed to provide is a `paymentAddress`. The following properties describ ...
```
yea it comes from `dcp-worker/etc/dcp-worker-config.js`

KirillKirnichansky committed:
README correction

bryan-hoang commented on package.json:
**suggestion:** Given that `npm` v6 is past its EOL, we should consider updating the minimum required version to use the package. 

e.g., add an `engines` field along with an `.npmrc` file 

```ini
# Disallow incorrect node or npm versions.
engine-strict=true
```

Naturally, this would also involve updating the `package-lock.json` file. We'll likely want to update it to `lockfileVersion` 3 (npm v9) so that Eddie can more easily deal with publishing.

```json
{
  "packageManager": "npm@9.6.7",
  "engines": {
    "node": ">=16",
    "npm": ">=7"
  }
}
```

bryan-hoang replied:
Applied in #5.

bryan-hoang commented on lib/useDCPWorker.js:
**question:** Can we finally stop tracking the `lib` folder through git? It adds noise during reviews and makes annoyingly dirty git work trees while developing locally.

I believe Eddie already makes sure the build command is ran before publishing so we shouldn't need to worry about keeping it tracked.

bryan-hoang replied:
Applied in #5.

bryan-hoang commented on README.md:
```diff
@@ -56,14 +52,14 @@ function Worker() {
-    - `fetchArguments: []`: A list of argument datum URIs that are safe to communicate with.
-    - `fetchData: []`: A list of input datum URIs that are safe to communicate with.
-    - `sendResults: []`: A list of URIs that are safe to send job results to.
+  - `allowOrigins?: object`: Allow list permitting network access beyond DCP messages to services. This list is used only in setting up the DCP Worker. After the worker is constructed/loaded, the `originManager` is responsible for managing origins (see Managing Origins). Empty by default of course.
```
**suggestion:** Remove wording/tone that could be interpreted as condescending. 

It could be not obvious to some folks. Implying that it's a trivial thing to assume isn't super welcoming.

```suggestion
  - `allowOrigins?: object`: Allow list permitting network access beyond DCP messages to services. This list is used only in setting up the DCP Worker. After the worker is constructed/loaded, the `originManager` is responsible for managing origins (see Managing Origins). It's empty by default.
```

bryan-hoang committed:
chore(package.json): sort for consistency

bryan-hoang committed:
build(npm): mandate minimum node & npm versions

With npm v6 at the end of its life, we need to update the minimum node
and npm versions the library supports. The lock file was updated with
npm v9 to minimize the amount of issues dealing with lock file version
incompatibilities.

bryan-hoang committed:
chore(lib): remove build output files

The files are auto generated when the project is installed and
published, so we no longer need to keep track of them through git.

bryan-hoang committed:
build(npm): add react{,-dom} as peer dependencies

Without it, build issues occur for applications running react 18. e.g.,
errors about being unable to import `react/jsx-runtime` occur.

bryan-hoang committed:
chore: git ignore common Node.js project files

While thinking about ignoring `*.tgz` files to make it less annoying to
use `npm pack`, I decided to add the Node.js template from
https://gitignore.io

bryan-hoang committed:
chore(package.json): update urls & license

KirillKirnichansky committed:
fix broken workflow for incorrect options address

Providing an incorrect paymentAddress in the workerOptions param would just break the hook and leave it in a broken state. There is a case where the param may not be defined on the first render/execution of the hook, but will be defined in the next render or a later time. The hook should be able to retry building the worker on subsequent rerenders.

Introduced the optionsError flag to capture when options validation and/or
construction fails before constructing the worker. Added handling of
optionsError to retry constructing the workerOptions obj and preventing worker intialization.

There were other React nature bugs that were resolved (useCallback
dependencies)

Also deprecated the options prop on workerStatistics since those values
can be found on the worker.

KirillKirnichansky committed:
fix: ensurePaymentAddress correctness

ensurePaymentAddress was incorrectly handling the case where
workerOptions.paymentAddress prop is defined, but, the value is
undefined. Realized that the check was faulty in general, so its been
corrected + better error info

KirillKirnichansky committed:
README improvements + how it works section

KirillKirnichansky committed:
remove unecessary types

originally wanted to remove OriginAccessManager types. Ended up removing
a bunch since some of the Worker props have changes which led to me
thinking its unecessary to always maintain the types of all DCP types
here. So is there value to having all of them defined? Or can I get away
with the any type

KirillKirnichansky committed:
remove workerOptions from return obj - this is accessible through worker

+ according README update

bryan-hoang commented on README.md:
```diff
@@ -56,51 +51,88 @@ function Worker() {
 - `identity?: Keystore`: A Keystore object (`dcp.wallet.Keystore`) which is passed to the Worker constructor and set as the Worker's identity when communicating over the network. If a Keystore is not provided, an arbitrary one will be generated.
 - `useLocalStorage?: boolean = true`:  A flag to toggle the use of the browser's local storage. The `workerOptions` object is the entity to be saved to local storage and is updated accordingly when calling `setWorkerOptions`.
-- `workerOptions: object`: This object is supplied to the Worker constructor as the `workerOptions` parameter (required). The only required property of the `workerOptions` object needed to provide is a `paymentAddress`. The rest of the properties will get default values.
+- `workerOptions: object`: The contents of this object will override the default values coming from the worker configuration coming from `dcpConfig`(provided by `dcp-client`, worker node sourced from `dcp-worker`). The only required property of the `workerOptions` object needed to provide is a `paymentAddress`. The following properties describe the worker options object configuring the DCP Worker ...
```
**suggestion:** Clarify the requirement of the `paymentAddress` option.

**nitpick:** Prefer using `:` rather than `.` when starting lists.

```suggestion
- `workerOptions: object`: The contents of this object will override the default values coming from the worker configuration coming from `dcpConfig`(provided by `dcp-client`, worker node sourced from `dcp-worker`). The only required property of `workerOptions` is `paymentAddress`. The following properties describe the worker options object configuring the DCP Worker:
```

bryan-hoang commented on README.md:
```diff
@@ -56,51 +51,88 @@ function Worker() {
   - `jobAddresses?: []`: If populated, worker will only fetch slices from jobs corresponding to the job addresses in this list.
   - `maxWorkingSandboxes?: integer | undefined`: Maximum number of sandboxes allowed to do work. If `undefined`, then the Supervisor will determine a safe limit, based off of machine hardware.
-  - `paymentAddress: Keystore | Address | String`: A Keystore or Address (`dcp.wallet.Address`) identifying a DCP Bank Account to deposit earned DCCs. An address string can also be supplied.
+  - `paymentAddress: Keystore | Address | String`: A Keystore, Address (`dcp.wallet.Address`) or string identifying a DCP Bank Account to deposit earned DCCs.
```
**suggestion:** I think the mandatory property should be listed first so that folks looking for the bare minimum can more easily parse it from skimming the text.

**suggestion:** I there's a difference between `String` and `string` given how `new String()` and `String()` work. Given that you're also using `boolean` rather than `Boolean`, we should opt for the more intuitive representation of the type.

```suggestion
  - `paymentAddress: Keystore | Address | string`: A Keystore, Address (`dcp.wallet.Address`) or string identifying a DCP Bank Account to deposit earned DCCs.
```

KirillKirnichansky replied:
Ive updated it so its: string | Address | Keystore. I believe string is preferable over String

bryan-hoang replied:
https://github.com/Distributive-Network/use-dcp-worker/pull/2#discussion_r1293642571

> I'd also like to fully address the **first half** of https://github.com/Distributive-Network/use-dcp-worker/pull/2#discussion_r1245730062 as well..

My bad, I wasn't being clear. I was referring to this part of the suggestion:

> suggestion: I think the mandatory property should be listed first so that folks looking for the bare minimum can more easily parse it from skimming the text.

i.e., Order `paymentAddress` first in the docs, rather than near the end/mixed in with the other parameters.

**question:** Should I split my suggestions in to multiple comments to help avoid this confusion in the future?

KirillKirnichansky replied:
> Should I split my suggestions in to multiple comments to help avoid this confusion in the future?

No this is my fault I had misinterpreted what your comment was - Sorry! Ill make this change.

bryan-hoang commented on README.md:
```diff
@@ -56,51 +51,88 @@ function Worker() {
+  - `error: Error | boolean`: Set when a worker error has occured, false otherwise.
+  - `workingSandboxes: integer`: Number of sandboxes currently doing work.
+- `workerStatistics: object`: Stores a global count of worker statistics for a browser session. Stored globally and preseved between component updates.
   - `slices: number`: Number of slices completed.
```
**suggestion:** `integer` isn't a valid primitive type in Typescript, but I get why it's described this way.

In other places, we use `number` even though `integer` could also work. For consistency and what I've normally seen, `change all instances of `integer` to `number`.

```suggestion
  - `workingSandboxes: number`: Number of sandboxes currently doing work.
- `workerStatistics: object`: Stores a global count of worker statistics for a browser session. Stored globally and preseved between component updates.
  - `slices: number`: Number of slices completed.
```

bryan-hoang commented on README.md:
```diff
@@ -56,51 +51,88 @@ function Worker() {
+  - workerOptions source always coming from dcpConfig
+  - added delay between quick worker fetching states
+  - Quality of life + maintainability improvements
+- __1.1.0__ - May
```
**suggestion:** This update is a major semver change as they break existing apps if people simply update their version and don't do anything else.

e.g., They'll need to migrate their usage of the hook from v1 to v2.

```suggestion
- __2.0.0__ - May
```

The version in `package.json` should also be changed.

bryan-hoang commented on src/useDCPWorker.tsx:
```diff
@@ -355,249 +291,265 @@ function getWorkerOptionsKey() {
+  if (Object.prototype.hasOwnProperty.call(loadedOptions, 'paymentAddress'))
+  {
+    if (loadedOptions.paymentAddress instanceof window.dcp.wallet.Keystore)
+    {
       loadedOptions.paymentAddress = new window.dcp.wallet.Address(loadedOptions.paymentAddress.address);
     }
-    else if (!(loadedOptions.paymentAddress instanceof window.dcp.wallet.Address)) {
+    else if (!(loadedOptions.paymentAddress instanceof window.dcp.wallet.Address))
+    {
       loadedOptions.paymentAddress = new window.dcp.wallet.Address(loadedOptions.paymentAddress);
     }
```
**question:** Why don't we pass the value to the constructor in either case?

The constructor accepts an instance of Address, so we can remove an unnecessary elseif statement.

bryan-hoang replied:
Scratch that, I didn't notice that the initial check was for a Keystore. Resolved.

bryan-hoang commented on src/useDCPWorker.tsx:
```diff
@@ -355,249 +291,265 @@ function getWorkerOptionsKey() {
-  ) {
+  }
+  else if (Object.prototype.hasOwnProperty.call(storage, 'defaultMaxWorkers'))
+  {
     loadedOptions = storage;
   }
 
-  if (!loadedOptions) return null;
+  if (!loadedOptions)
+    return null;
 
-  if (
-    Object.prototype.hasOwnProperty.call(loadedOptions, 'paymentAddress')
-  ) {
-    if (loadedOptions.paymentAddress instanceof window.dcp.wallet.Keystore) {
+  if (Object.prototype.hasOwnProperty.call(loadedOptions, 'paymentAddress'))
```
**suggestion:** You may want to look into [`Object.hasOwn`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn) to cut down on the verbosity ([added 2021/2022](https://github.com/tc39/proposal-accessible-object-hasownproperty)).

**praise:** Kudos on not falling back to simply calling `hasOwnProperty` directly!

bryan-hoang replied:
**suggestion:** `src/useDCPWorker.tsx:498` contains the remaining uses of `hasOwnProperty` which should be replaced with `Object.hasOwn`.

KirillKirnichansky committed:
README corrections

corrections suggested by Bryan in MR

KirillKirnichansky committed:
version 2 in package.json

KirillKirnichansky committed:
use Object.hasOwn instead of verbose alternative

updated Object.prototype.hasOwnProperty.call to Object.hasOwn

KirillKirnichansky committed:
fix: add es2022 lib to tsconfig

needed to support Object.hasOwn

KirillKirnichansky committed:
replace maxWorkingSandboxes with cores configuration

new supervisor in worker API deprecates the maxWorkingSandboxes concept
and replaces it with cores. maxWorkingSandboxes is still supported but
will eventually be completely deprecated. Similarily, this commit
supports the new cores property as maxWorkingSandboxes was supported,
while still keeping the latter

STILL WIP - as of testing, proxy logic seems broken (not saving cores to
local storage)

KirillKirnichansky committed:
applyWorkerOptions does proper leaf merge, workerOptions proxy handles
nested objects and better handling of workerOptions.cores.cpu

- applyWorkerOptions now does a full leafMerge, using leafMerge from
dcp/utils library, whilst preserving reference to dcpConfig.worker.
- workerOptions proxy setter now handles for nested properties (to
accomodate changes to workerOptions.cores.cpu)
	- can now save cores.cpu to localstorage
- remove debug logs and debuggers

KirillKirnichansky committed:
handle case where cores and maxWorkingSandboxes defined

specifically before the worker constructor
this can happen if maxWorkingSandboxes and cores are set in
workerOptions before it is passed to the Worker constructor

KirillKirnichansky committed:
fix newly introduced hasOwnProperty to hasOwn

bryan-hoang commented:
@KirillKirnichansky Heads up, I fixed up the commit bumping the package version so that the change was also reflected in the `package-lock.json` file.

bryan-hoang commented on README.md:
```diff
@@ -56,51 +51,88 @@ function Worker() {
+- `remove_byKey(key):` Removes all origins for all purposes under the _key_. A `null` _key_ is not accepted, must be a string.
+- `getAllowList(purpose, key):` returns a list of origins under the _purpose_ and _key_. A `null` _purpose_ is not accepted, must be a string. Previously added origins under `null` _purpose_ and/or _key_ will match any _purpose_ and/or _key_, respectively.
+- `isAllowed(origin, purpose, key):` returns `true` if _origin_ is allowed under the _purpose_ and _key_. A `null` _purpose_ is not accepted, must be a string. Previously added origins under `null` _purpose_ and/or _key_ will match any _purpose_ and/or _key_, respectively.
+# Change log
+- __1.0.0__ - Feb 28, 2023
+  - Inital Release
+- __1.0.1__ - Mar 1, 2023
+  - Added src/ directory to published files to support source map
+- __1.0.2__ - Mar 23, 2023
+  - local storage will only save paymentAddress and maxWorkingSandbox props
+  - workerOptions source always coming from dcpConfig
+  - added delay between quick worker fetching states
+  - Quality of life + maintainability improvements
+- __2.0.0__ - May
+  - now returns the worker
+  - proper handling of race-condition in constructing the worker when/if the hook is executed multiple times at once
+  - workerOptions passed to worker constructor is a Proxy now
+    - needed reflect changes to paymentAddress and maxWorkingSandbox props to local storage and trigger component update/re-render
+  - removed workerState.isLoaded since now we are returning the worker
+    - !isLoaded === !worker
+  - removed workerOptionsState, start/stop/toggleWorker, sandboxes, originManager
+    - these are all accessible via Worker API
+  - removed applyWorkerOptions since editing workerOptions should be done by directly mutating worker.workerOptions
+  - improved error handling
+  - quality of life improvements
```
**suggestion:** Order the Changelog in reverse chronological order.

Most folks want to see the latest changes described first [^1]. e.g., On a new release, they'll want to see the new changes listed first in the file, not at the end of it after scrolling down.

[^1]: https://keepachangelog.com/en/1.0.0/#how

bryan-hoang commented on src/useDCPWorker.tsx:
```diff
@@ -71,81 +45,21 @@ declare interface IWorkerOptions {
+  };
   maxWorkingSandboxes?: number | undefined;
-  paymentAddress?: Address | string | null;
+  paymentAddress?: any;
```
**question:** Why loosen the type?

i.e., were there issues with `Address | string | null`?

KirillKirnichansky replied:
I ended up removing a bunch of the types, including `Address`, since it felt like more maintenance to ensure the types were up to date. Since the `Worker`, `Supervisor` and events changed in the `dcp-client` library I had to also reflect those changes here. In the end I thought having the types weren't providing much value, so I made it easier and removed them (perhaps this was a lazy decision). Ill bring back `Address` in this case as that type isn't subject to change much, if at all ...

bryan-hoang replied:
**thought:** In an ideal world, we ship types with the `dcp-client` package, making it easier to update all the types while keeping up to date with the actual runtime changes being made to the code. Currently, `@types/dcp-client` only supports types on the Node.js side. It'd be fantastic if it also supported the types on the Web side as well, but that's out of scope for this PR.

> it felt like more maintenance to ensure the types were up to date.

Fair point. It sounds like we shouldn't be using TypeScript at all then if we're not willing to pay the burden as library maintainers to make better developer experiences for application developers. But doing it right the first time is less work overall than having to deal with folks filing issues down the line about invalid/incomplete types, or asking for better types from the library. There are quite a few folks who base their library choices based on TS interop/support from what I've seen.

bryan-hoang commented on src/useDCPWorker.tsx:
```diff
@@ -227,20 +166,26 @@ const defaultWorkerStats: IDefaultWorkerStats = {
-  if (action.type === 'INCREMENT_COMPUTE_TIME') {
+  if (action.type === WorkerStatsActions.ADD_COMPUTE_TIME) {
     updatedStats.computeTime += action.data;
-  } else if (action.type === 'INCREMENT_SLICES') {
+  } else if (action.type === WorkerStatsActions.ADD_SLICE) {
     updatedStats.slices++;
-  } else if (action.type === 'INCREMENT_CREDITS') {
+  } else if (action.type === WorkerStatsActions.ADD_CREDITS) {
     updatedStats.credits = updatedStats.credits.plus(action.data);
   }
```
**nitpick:** If other braces will be changed to Allman style, these braces should be changed as well.

KirillKirnichansky replied:
fixed to use switch case instead

bryan-hoang commented on src/useDCPWorker.tsx:
```diff
@@ -355,249 +295,277 @@ function getWorkerOptionsKey() {
+   *  This method ensures that the paymentAddress prop in workerOptions is of valid type.
+   *  @returns `true` if valid, `false` otherwise
+   */
+  const ensurePaymentAddressType = useCallback(() => {
```
**issue:** If I don't set the payment address, there's currently no way at runtime to handle the situation.

Why don't we throw an error on the invalid option being passed in? Or at the very least, return the error somehow so that consumers can decide what to do in this situation.

We throw an error when `window.dcp` isn't available. The lack of an address should be treated in a similar manner imo.

KirillKirnichansky replied:
Currently the error is logged to the console. Currently if an invalid or missing paymentAddress is provided then the hook wont pass the constructWorkerOptions stage. However, if the the value for paymentAddress gets defined/corrected during a re-render, then the hook will re-attempt constructing. In the case where using `useState` for the value of paymentAddress in the hook params, the state may not be defined with a valid value by default. In that case, if I was to throw, then the app would not be in a recoverable state. Similarly, `ensurePaymentAddressType` is called on changes to the worker.workerOptions.paymentAddress - I don't want to throw in those cases either.

bryan-hoang replied:
Ah, okay. I didn't grok that from reading the code. 

**suggestion:** Document the hook's behaviour when an invalid address is passed. i.e., describe what a consumer would need to do to handle that situation through the use of `isLoaded`. Because it's not clear to me how one should handle an erroneous condition from just reading the code/docs.

I'd also like to fully address the first half of https://github.com/Distributive-Network/use-dcp-worker/pull/2#discussion_r1245730062 as well, since it's still unclear from skimming the docs which options are mandatory or not. I know it's implied from the example, but making it more obvious in the reference docs would be great!

KirillKirnichansky replied:
I added a note under the paymentAddress param desc

bryan-hoang replied:
Thanks for adding the docs!

> I'd also like to fully address the first half of https://github.com/Distributive-Network/use-dcp-worker/pull/2#discussion_r1245730062 as well...

Could you please implement that comment as well, or at least say why it's not worth addressing?

KirillKirnichansky replied:
My apologies I didnt mention that I had addressed it already

KirillKirnichansky committed:
use switch case instead of if/else chain

KirillKirnichansky committed:
bring back Keystore and Address types for paymentAddress

KirillKirnichansky committed:
make README changelog reverse chronological + add 1.0.4 log

KirillKirnichansky committed:
Improve paymentAddress param documentation

better ordering of types + describe behaviour when passing invalid
paymentAddress

KirillKirnichansky committed:
order paymentAddress first in README as the mandatory param

bryan-hoang reviewed (approved):
LGTM!
